src/Features/Core/Portable/TodoComments/AbstractTodoCommentService.cs

```diff
@@ -42,9 +55,32 @@ internal abstract class AbstractTodoCommentService : ITodoCommentService
         private async Task<IList<TodoComment>> GetTodoCommentsInRemoteHostAsync(
             RemoteHostClient client, Document document, ImmutableArray<TodoCommentDescriptor> commentDescriptors, CancellationToken cancellationToken)
         {
-            return await client.RunCodeAnalysisServiceOnRemoteHostAsync<IList<TodoComment>>(
-                document.Project.Solution, nameof(IRemoteTodoCommentService.GetTodoCommentsAsync),
-                new object[] { document.Id, commentDescriptors }, cancellationToken).ConfigureAwait(false);
+            var keepAliveSession = await GetKeepAliveSessionAsync(client).ConfigureAwait(false);
+
+            var (success, result) = await keepAliveSession.TryInvokeAsync<IList<TodoComment>>(
+                nameof(IRemoteTodoCommentService.GetTodoCommentsAsync),
+                document.Project.Solution,
+                new object[] { document.Id, commentDescriptors }).ConfigureAwait(false);
+
+            return success ? result : SpecializedCollections.EmptyList<TodoComment>();
+        }
+
+        private async Task<KeepAliveSession> GetKeepAliveSessionAsync(RemoteHostClient client)
+        {
+            if (_session != null)
+            {
+                return _session;
+            }
+
+            var session = await client.TryCreateServiceKeepAliveSessionAsync(WellKnownServiceHubServices.CodeAnalysisService, CancellationToken.None).ConfigureAwait(false);
+            var oldSession = Interlocked.CompareExchange(ref _session, session, null);
+            if (oldSession != null)
+            {
+                // there was race, kill the lost session
```

Comment:
This comment still applies.  I don't see why we keep using Interlocked all over hte place when it makes things much more complex and much harder to vet for concurrency issues.